### ISSUE TITLE
eslint throws exception when no configuration file found.

### ISSUE THREAD

**Issue by cowboyd** [eslint/eslint#7144]
When no configuration file is present, eslint prints an exception backtrace to the console and returns an exit status of `1`. This throws a spanner in the works of automated syntax checkers like flycheck [as seen here][1].

```
$ echo 'var x = "y";' | eslint --stdin                                                                                                                                                                
No ESLint configuration found.
Error: No ESLint configuration found.
    at getLocalConfig (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/lib/config.js:162:35)
    at Config.getConfig (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/lib/config.js:256:22)
    at processText (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/lib/cli-engine.js:222:33)
    at CLIEngine.executeOnText (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/lib/cli-engine.js:749:26)
    at Object.execute (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/lib/cli.js:174:42)
    at /Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/bin/eslint.js:57:36
    at ConcatStream.<anonymous> (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/node_modules/concat-stream/index.js:36:43)
    at emitNone (events.js:72:20)
    at ConcatStream.emit (events.js:166:7)
    at finishMaybe (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/node_modules/concat-stream/node_modules/readable-stream/lib/_stream_writable.js:475:14)
```

One possible solution would be to make this a warning not an error, or also to have an option to suppress the error / warning. e.g. `eslint --no-configuration-ok`

### Environment

```
> eslint "--version"
>v3.5.0
> node -v                                                                                                                                                   
> v4.2.3
> npm -v
```

[1]: https://github.com/flycheck/flycheck/issues/1085

**Comment by platinumazure** [comment]
This was an intentional change with ESLint 3.0.0.

When ESLint doesn't have a configuration file and no .eslintrc.* file is found in the CWD, ancestor directories, or in a user's home directory, ESLint used to run with no rules turned on (so that it only parsed). That wasn't very useful to a lot of people, so we decided to require either a configuration file, or a CLI option that informed ESLint about what rules to run or not run.

Please see the [section in the migration guide](http://eslint.org/docs/user-guide/migrating-to-3.0.0#requiring-configuration-to-run) for more information.

**Comment by Simplify** [comment]
Just for the record:
flycheck prefers structured output and we use --format=checkstyle. We are expecting result as XML but we get text output:

```
$ eslint --format=checkstyle ember-cli-build.js 

Oops! Something went wrong! :(

ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:

    eslint --init

ESLint looked for configuration files in /Users/sasa/Projects/toptal/frontend and its ancestors.

If you think you already have a configuration file or if you need more help, please stop by the ESLint chat room: https://gitter.im/eslint/eslint
```

is it possible to format this error as CheckStyle XML for first line and first character in file? In that case error will be shown on the first line in Emacs buffer.

**Comment by platinumazure** [comment]
@Simplify That's something we can consider. However, I want to make sure you're aware of another workaround you can already use: If flycheck is a Node app and/or if you're willing to write a simple Node app, you could use our [CLIEngine API](http://eslint.org/docs/developer-guide/nodejs-api) to run ESLint, and the error will be merely a thrown exception instead of CLI output. Then you could put a try/catch around it and handle it using CheckStyle XML (or any other way) if you like.

And, of course, you also might consider ensuring the presence of an ESLint config file before calling ESLint. :smile:

**Comment by cowboyd** [comment]
@platinumazure flycheck is an elisp package, so would prefer not to have to install any extra supporting node packages into the environment just to support an editor.

The problem isn't so much that flycheck is erroring, but that it's impossible to distinguish the error from system level errors like not being able to resolve a dependent npm package. Kinda like an API that returns a `500` for a validation error.

**Comment by Simplify** [comment]
@platinumazure [Flycheck](http://www.flycheck.org/en/latest/) is GNU Emacs extension written in Emacs lisp. It uses various syntax checking and linting tools to automatically check the contents of buffers while you type, and reports warnings and errors directly in the buffer, or in an optional error list. Our only option is to eslint executable.

**Comment by platinumazure** [comment]
@Simplify Thanks for explaining about your execution environment, that's helpful.

@eslint/eslint-team So we have two proposals for the problem described in this issue. One, provide a CLI option `--no-configuration-okay` or similar which disables the configuration check. Two, allow uncaught exceptions to be somehow output through the user's desired formatter. Can we consider either of these? Is there a better option?

**Comment by Simplify** [comment]
@platinumazure Please don't do `--no-configuration-okay`, there is already `--no-eslintrc` but we can not use it. The reason is that we can not detect that ESLint configuration is present, due to 6 different ways to configure ESLint. With most checkers/linters we can detect configuration up to users home directory, because most tools use same name for config file. With ESLint is that not possible, because there are 5 ways to name config file and there is also option to place configuration in `package.json`. `--no-configuration-okay` can be useful only if ESLint still try to find configuration and use it if found.

The best way to solve this issue is respect `--format` option given to `eslint` executable, not only for flycheck, but also for any other tool that expect structured output from ESLint.

**Comment by platinumazure** [comment]
@Simplify I added a third proposal in my earlier post (edited in later, so you may have missed it). That proposal is to use a different exit code for failed linting (which should output using the formatter) vs uncaught exception (in which all bets are off). Right now, both have an exit code of 1. If an uncaught exception had a different exit code, would that be sufficient to tell flycheck not to try to parse the output as Checkstyle XML?

**Comment by Simplify** [comment]
@platinumazure No, don't think so, we always expect structured output. 

@lunaryorn Any thoughts?

**Comment by lunaryorn** [comment]
@cowboyd @Simplify Thank you for bringing this issue up.

@platinumazure As a maintainer of Flycheck—i.e. a tool that runs eslint in an automated, noninteractive fashion—I'd prefer if it did not error out at all.

As things stand eslint is a lot less useful to us than it could be.  As it refuses to do anything in the absence of a configuration file we can't use it as default syntax checker and linter for Javascript files: Without a configuration it doesn't provide any valuable feedback at all, not even about simple syntax errors—which already is a lot for many users.  We can't even gracefully escalate to eslint because there's no simple way for us to check whether it can find a configuration somewhere—there's no option like `--check-config` or so.

I'm not sure why you think that eslint without a configuration would not be useful… even a simple syntax check provides valuable feedback to the user, doesn't it?

I understand that you'd like to make users aware of the potential in configuring eslint, but a less invasive way would serve us better, like a warning emitted before the regular output of a syntax checker.

----

If it's altogether impossible to go for a less invasive way my second best choice would be if eslint did not error out with an exception but reported a regular error, in the chosen output format.  That'd still not let us pick eslint as default but at least the absence of a configuration wouldn't make it blow up anymore.

**Comment by Simplify** [comment]
Actually, having `--check-config` or '--show-config-location` option in ESLint that prints location of configuration  file of showing notice that there is no configuration is best solution to this problem. In the case that there is configuration, we can still be using ESLint as default JS checker in Emacs and keep ESLint as first JS checker. If there is no configuration we know that using ESLint can not be used. In that case we can fallback to JSHint or something else.

@platinumazure, is implementing this an option for ESLint team?

**Comment by nzakas** [comment]
Can someone please summarize this thread? I'm having a hard time seeing a clear definition of the problem and the three proposals to consider.

One thing I do want to point out: we don't have a standard output format for non-code errors. Our formatters are designed specifically to output linting/parsing errors, or other errors that are found in the code. Our general approach is that all other errors throw.

**Comment by Simplify** [comment]
@nzakas let me summarize this problem:

1. Flycheck is Emacs extension. It uses various syntax checking and linting tools to automatically check the contents of buffers while you type, and reports warnings and errors directly in the buffer, or in an optional error list.
2. Flycheck supports around 80 different "checkers" for many different languages and ESLint is one of them. There are also extensions for Flycheck that add more than 20 more "checkers". Flycheck is also one of the 10 most downloaded Emacs extensions according to [Melpa](https://melpa.org/).
3. Out default JS "checker" is ESLint doe the quality it provides (kudos for that).
4. One of our goals is to provide **best user experience possible** for our user. That means that when somebody opens JS file with GNU Emacs we can show them all problems in that file.
5. In older versions of ESLint that was possible. 
6. From ESLint 3.0, eslint "checker" will not display any information without configuration file. I understand your decision to not use any "default" configuration. We didn't saw this problem until today. I personally have ESLint configuration in home directory and I was not affected.
7. If some "checker" needs configuration file in order to work, we have build in solution to detect if configuration exists, usually searching for config file by name. In the case of ESLint that is impossible, there are 5 different names for configuration file and there is possibility that configuration sits in `package.json`.
8. If we can't run default "checker" (ESLint in this case) we run some other alternative (JSHint for example).
9. Right now we can not determinate if we can run ESLint. Even when we detect executable, we can not detect of configuration exists. However, we'll like to keep ESLint as default checker.
10. Best solution for this problem is that we "ask" ESLint if configuration exists. Then we can make decision to enable/disable ESLint in Emacs buffer. Something like `--check-config` or `--show-config-location` is probably best way to do it.

**Comment by platinumazure** [comment]
@Simplify Just out of curiosity, what happens when you run `eslint --print-config` when no config is available? Wondering if that could give a clue about presence or absence of config, but it wouldn't be very useful if we populate some defaults in that case.

**Comment by Simplify** [comment]
I just try it and now I'm confused. Without config I get same output as when running `eslint --help`. With config `.esintrc.js` file in same directory (just generated with `--init`) I also get same output. A bug?

```
$ eslint -v
v3.5.0
```

Config is present:
```
$ ls -la 
total 128
<some files skipped>
-rw-r--r--    1 sasa  staff   7355 Sep 13 23:19 .eslintrc.js
<more files skipped>
```

Running with `--print-config` option:
```
$ eslint --print-config
eslint [options] file.js [file.js] [dir]

Basic configuration:
  -c, --config path::String   Use configuration from this file or shareable config
  --no-eslintrc               Disable use of configuration from .eslintrc
  --env [String]              Specify environments
  --ext [String]              Specify JavaScript file extensions - default: .js
  --global [String]           Define global variables
  --parser String             Specify the parser to be used
  --parser-options Object     Specify parser options
<some lines skipped>
Miscellaneous:
  --init                      Run config initialization wizard - default: false
  --fix                       Automatically fix problems
  --debug                     Output debugging information
  -h, --help                  Show help
  -v, --version               Output the version number
  --no-inline-config          Prevent comments from changing config or rules
  --print-config              Print the configuration to be used
```

**Comment by platinumazure** [comment]
Ack, my fault. You have to specify a file (`eslint --print-config some-file.js`).

**Comment by Simplify** [comment]
Ok, `--print-config String` in help will be more helpful.

With config file you get JSON configuration printed out, but when running without config, exit status is still 1 and same error message is displayed. Why just not print "ESLint can't find any configuration" or empty JSON `{}` when running with `--print-config` at least? 

@lunaryorn We can try to use this exit status as workaround in `:predicate` with `eslint --print-config source`. is that ok with you?

**Comment by platinumazure** [comment]
Thanks for trying that out, I agree that experience can be improved.

Let's see what the team thinks about your summarized proposal.

**Comment by Simplify** [comment]
Well, out `predicate` runs every time before Emacs buffer is checked, so running `eslint --print-config <filename>` every time user changes buffer is not really an option.

**Comment by lunaryorn** [comment]
@Simplify I think we can handle that.  I believe that `--print-config` is sufficient for our purposes, and I think we should rather improve the way we filter checkers…

@platinumazure Thanks a lot for your help!

**Comment by nzakas** [comment]
Does that mean we don't need to make any changes?

**Comment by platinumazure** [comment]
@nzakas I can't speak to whether we need to make changes for flycheck's issues, but I definitely will write a PR to improve the CLI documentation around `--print-config`. (I might actually restructure the option to take a file path string argument, which will require tweaking some of the code in `bin/eslint.js` as well since right now it looks for a positional file name argument when running in "print config" mode.)

I also think it might be worth revisiting the discussion around exit codes (use 1 for lint failure, something else for exceptions), but it doesn't seem like that will help the flycheck folks much.

I will let them speak to whatever changes they might need for us to do.

**Comment by nzakas** [comment]
@platinumazure sorry for being unclear, I was asking @lunaryorn.

I don't think it's worth messing with exit codes. The last time we tried to change exit codes (make the exit code the number of lint errors), it was a disaster and broke a lot of people's CI systems. I don't think changing exit codes is useful enough to consider it again.

In the meantime, this is still labeled as a question. If it might result in a change, please update the labels.

**Comment by lunaryorn** [comment]
@nzakas Let's say that we can cope with eslint's current behaviour: `--print-config` provides sufficient information to detect whether eslint would work for a particular file which—with some additional infrastructure in our code—provides what we need to make a decision about whether we should use eslint for a file.

So if you're asking me whether you can get away without making any changes to eslint I should say, yes, you can.  We can solve the issue at hand on our side.

But if you're asking me whether you should make changes to eslint's interface my answer would be different.  I think that eslint's current interface is unfortunate in many different ways, but I understand that I'm not in a position to argue.

**Comment by mysticatea** [comment]
> flycheck prefers structured output and we use --format=checkstyle. We are expecting result as XML but we get text output:
> https://github.com/eslint/eslint/issues/7144#issuecomment-246710515

I'm interested in this.
Our formatter feature cannot format error messages, but tools would expect formatted error messages.
It sounds reasonable to me if we add the feature to format error messages into formatters.

For example:

```js
// formatters/json.js

module.exports = function(results) {
    return JSON.stringify(results);
};

module.exports.formatError = function(err) {
    return JSON.stringify([{
        fatal: true,
        message: err.detailMessage,
    }]);
};
```

- If the `formatError` function is not defined, ESLint shows errors as same as current.
- Node.js API does not use this feature since someone can use `try-catch`.
- CLI uses this feature to show errors in the specific format.
- `err` is an instance of `Error` constructor.
- We have [message templates for critical errors](https://github.com/eslint/eslint/tree/master/messages). `err.detailMessage` has the message.

**Comment by nzakas** [comment]
@mysticatea I disagree. I don't think it makes sense to have formatters also format errors. We specifically handle fatal errors differently so we can provide better console output. Letting formatters into that process not only complicates things, but also could make the output harder to read for humans.

**Comment by nzakas** [comment]
@lunaryorn what is unfortunate about the interface? AFAIK ESLint works the same as most command line tools with output formatting: expected results are formatted, unexpected errors crash. 

Your comment made it sound like you had figured out how to solve your problem, but if not, feel free to share what you think a good solution would be.

**Comment by platinumazure** [comment]
@nzakas Most CLI tools differentiate between normal failure output and fatal crash output by exit code, as well. As I said before, I think we should exit 2 for crashes.

**Comment by lunaryorn** [comment]
@nzakas Yes we can solve our problems in Flycheck, and already in the process of doing so.  That said, I still disagree with much of what you said.

Is a missing configuration unexpected?  Let alone, a fatal crash?  I think not.  But then why throw a stracktrace at a user who just made a mistake that any user could make?  Might be just me but to me feels like being shown the "f*** you I don't care" finger.  Just imagine a JS beginner seeing that message…

And then, what's the purpose of JSON and XML formats?  Integration with other programs.  But what good is that integration if you give it up for errors?

Of course it'd be harder for a human to read errors in structured output but who reads structured output in the first place?  Do you read your eslint output in XML or JSON?  Of course not.  

The XML is for Jenkins and friends to show the eslint results in a fancy UI—and what good is that integration if I've got to leave the fancy UI and dig the raw console log to figure out an eslint configuration mistake on my CI server?

That is my point of view, since you asked for it.   But I do not intend to restart a discussion you already had, nor to I have the interest or the energy to argue for change.  I'll be unsubscribing from this thread.  Please mention me if you have questions, but if possible keep me out of the discussion.

I'd like to thank you for your help on this issue, and for pointing us towards `--print-config` which solves our immediate issue.  And I'd also like to express my heartfelt thanks to you and all eslint contributors; eslint is an awesome and indispensable tool that's essential for my work and goes a long way to make JS programming at least half-way bearable to me.  Many, many thanks.

**Comment by cowboyd** [comment]
> I'd like to thank you for your help on this issue, and for pointing us towards --print-config which solves our immediate issue. And I'd also like to express my heartfelt thanks to you and all eslint contributors; eslint is an awesome and indispensable tool that's essential for my work and goes a long way to make JS programming at least half-way bearable to me. Many, many thanks.

Let me concur. Eslint is a fantastic tool, and I filed this issue in the hope to improve it. I am very happy with it now, but my on-boarding process was not as smooth as it perhaps could have been.

I don't know what the best solution is, but I can say as a brand new user (which I was), a missing configuration, be it invoked from the command line or from a tool, is almost _guaranteed_ to be the first thing you encounter, and it seems like punting on the UX around that use-case is leaving a sharp edge in place.

**Comment by nzakas** [comment]
@lunaryorn thanks for your feedback. I'd just like to answer some of your points, not to debate, but to explain the reasons why things are the way they are.

> Is a missing configuration unexpected? Let alone, a fatal crash? I think not. 

Yes, it is. This came after years of getting messages and bugs filed saying, "ESLint isn't doing anything." This just kept happening, so finally, we made it so you must have a configuration in order to run ESLint. You can see the discussion here: https://github.com/eslint/eslint/issues/5987. The bottom line is that we decided ESLint cannot be run without a config, so yes, it is considered fatal to be missing a configuration file.

> But then why throw a stracktrace at a user who just made a mistake that any user could make? Might be just me but to me feels like being shown the "f*** you I don't care" finger. Just imagine a JS beginner seeing that message…

I'll admit, it's pretty dumb to throw a stack trace, and in fact, this is a bug that we should fix (filed https://github.com/eslint/eslint/issues/7213). We actually do print out a friendly error message when passing files on the command line, but for some reason, it's not getting shown when you pipe in code. You can see the error message here: https://github.com/eslint/eslint/blob/master/messages/no-config-found.txt

> The XML is for Jenkins and friends to show the eslint results in a fancy UI—and what good is that integration if I've got to leave the fancy UI and dig the raw console log to figure out an eslint configuration mistake on my CI server?

In general, we expect users to only see a missing configuration error the first time they setup ESLint rather than on an ongoing basis inside of a CI process. That's why we (tried) to optimize the error message towards the user sitting at a console with the aforementioned pretty error message.

@cowboyd  regarding:

> I don't know what the best solution is, but I can say as a brand new user (which I was), a missing configuration, be it invoked from the command line or from a tool, is almost guaranteed to be the first thing you encounter, and it seems like punting on the UX around that use-case is leaving a sharp edge in place.

Please see my note above that this is, in fact, a bug that we will fix.

**Comment by vitorbal** [comment]
#7213 is merged and released as of v3.7.0. It also looks like all the questions brought up in this issue have been answered (apologies if that is not true, please correct me if I'm wrong). Is anyone opposed to closing this now?

**Comment by nzakas** [comment]
Nope, it seems like we've resolved what we can resolve.
